### PR TITLE
Add general quirk for canvas2d fingerprinting

### PR DESCRIPTION
**Pull request by sysrqb**
#### b680f873f60959c6077d9a9297ab933a35d7bbf4
<pre>
Add general quirk for canvas2d fingerprinting
<a href="https://bugs.webkit.org/show_bug.cgi?id=260842">https://bugs.webkit.org/show_bug.cgi?id=260842</a>
rdar://113039941

Reviewed by Wenson Hsieh.

In 266400@main I added a quirk for canvas fingerprinting that targeted specific
pages on two specific domains. As a follow up to that, this patch expands the
application of that quirk broadly by comparing attributes of the script instead
of looking at the specific document URL. This step is important because we know
protections against canvas fingerprinting cause breakage on some sites, and
there is likely more greater impact than we&apos;re aware at this time.

Similar to the current behavior, this patch relies on matching the last drawn
text string. In addition, it compares the dimensions of the canvas and the size
of the script. These characteristics seem sufficiently descriptive that we can
safely use them for identifying the target script.

This patch also introduces the CanvasFingerprintingQuirkEnabled preference for
controlling whether the quirk is available. It also adapts the existing API
test.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::toDataURL):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::advancedPrivacyProtectionSubstituteDataURLForScriptWithFeatures const):
(WebCore::Quirks::shouldEnableCanvas2DAdvancedPrivacyProtectionQuirk const): Deleted.
(WebCore::Quirks::advancedPrivacyProtectionSubstituteDataURLForText const): Deleted.
* Source/WebCore/page/Quirks.h:
* Source/WebCore/platform/Logging.h:
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/267513@main">https://commits.webkit.org/267513@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/34fa039f56a010890e699e6aaf1c38674ddb2e78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18459 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15636 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16868 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20231 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17139 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17941 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16876 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17257 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19244 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14501 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15113 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21883 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14393 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15490 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15279 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19584 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15905 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15886 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13483 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18235 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15074 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4273 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4028 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19440 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19456 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15715 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4118 "Passed tests") |

### CODE REVIEW

File: Source/WebCore/platform/Logging.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ namespace WebCore {
     M(EventRegions) \
     M(FileAPI) \
     M(Filters) \
+    M(Fingerprinting) \
```
Comment (whsieh) [inline Source/WebCore/platform/Logging.h]: Nit - maybe, `FingerprintingMitigations`?

File: Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm [hunk 2]
```diff
@@ -1009,14 +1009,15 @@ HTTPServer server({
 TEST(AdvancedPrivacyProtections, Canvas2DQuirks)
 {
     using namespace TestWebKitAPI;
-    const auto script = @"let canvas = document.createElement(\"canvas\"); let ctx = canvas.getContext(\"2d\"); ctx.fillText(\"<@nv45. F1n63r,Pr1n71n6!\", 10, 40); canvas.toDataURL();";
+    auto* script = @"let canvas = document.createElement(\"canvas\"); canvas.width = 280; canvas.height = 60; let ctx = canvas.getContext(\"2d\"); ctx.fillText(\"<@nv45. F1n63r,Pr1n71n6!\", 10, 40); canvas.toDataURL();";
```
Comment (whsieh) [inline Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm]:
```suggestion
    auto *script = @"let canvas = document.createElement(\"canvas\"); canvas.width = 280; canvas.height = 60; let ctx = canvas.getContext(\"2d\"); ctx.fillText(\"<@nv45. F1n63r,Pr1n71n6!\", 10, 40); canvas.toDataURL();";
```

File: Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm [hunk 3]
```diff
@@ -1009,14 +1009,15 @@ HTTPServer server({
...
+    auto* paddedScript = [script stringByPaddingToLength:212053 withString:@" " startingAtIndex:0];
```
Comment (whsieh) [inline Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm]:
```suggestion
    auto *paddedScript = [script stringByPaddingToLength:212053 withString:@" " startingAtIndex:0];
```

**Review by whsieh — APPROVED** [review]

File: Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm (hunk 2 above)
Comment (sysrqb) [reply Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm]: Looking at the current codebase, it seems like `auto*` is more common than `auto *`.

Comment (rniwa) [reply Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm]: Hm... I think we generally use `auto *` in Objective-C files whereas we use `auto* ` in C++ files.

Comment (cdumez) [reply Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm]: Yes, we use `Type *` instead of `Type*` in Objective C code.

Comment (sysrqb) [reply Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm]: Ah, thanks for clarifying

**Pull request merged**